### PR TITLE
Update tls logic to work better with guarded call

### PR DESCRIPTION
**Pull request by albanD**
Description of the new behavior is in PythonFallbackKernel.cpp.
The updated test makes sure that we only call alias on the first Tensor.

### CODE REVIEW

**Review by soulitzer — COMMENTED** [review]
Pretty cool! Just one small comment on naming:

I think some of the meaning of the terms shifted during this PR, which confused me for a bit. Previously TLSState referred to the current local dispatch key set. In this PR TLSState seems to refers to `tls_on_entry`. Stash also means something slightly different now.

**Comment by soulitzer** [comment]
Also TLSState = Thread Local State State? :P

File: test/test_autograd.py [hunk 1]
```diff
@@ -7954,10 +7957,11 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
         s = MySubclass(a)
 
         with fwAD.dual_level():
+            # Only the primal has "alias" called on it
             fwAD.make_dual(s, torch.rand_like(s))
             self.assertEqual(counter[0], 1)
             fwAD.make_dual(torch.rand_like(s), s)
-            self.assertEqual(counter[0], 2)
+            self.assertEqual(counter[0], 1)
```
Comment (soulitzer) [inline test/test_autograd.py:7964]: I think I realize what is happening here now, my test last time was completely oblivious to what torch function does.

Before this PR __torch_function__ was wrapping the output of torch.rand_like in the subclass, but now that we disable torch function for this subclass, that no longer happens, and the counter is only triggered once as expected.

File: aten/src/ATen/core/PythonFallbackKernel.cpp [hunk 2]
```diff
@@ -6,23 +6,56 @@
...
 namespace {
 
-// TLS saving the state of the include/exclude sets on entry to the dispatcher
-// This is set in the pythonTLSSnapshot fallback and used by the Python fallback.
-thread_local std::stack<c10::impl::LocalDispatchKeySet> tls_on_entry;
+// This TLS is used to track the state of the dispatcher to be able to restore
+// it when calling back into python.
+// It has the following invariant:
+//  - It must be empty while python code is executed.
+//  - It should only be set once even for multiple dispatcher calls that do not come
```
Comment (bdhirsh) [inline aten/src/ATen/core/PythonFallbackKernel.cpp:13]: and specifically, it gets set by the first dispatcher call, and ignored by any subsequent dispatcher calls? (until you eventually get to the python fallback and clear it).

File: test/test_autograd.py (hunk 1 above)
Comment (albanD) [reply test/test_autograd.py:7964]: Yep

File: test/test_autograd.py [hunk 3]
```diff
@@ -7939,13 +7939,16 @@ class MySubclass(torch.Tensor):
             def __new__(cls, data=None):
                 return torch.Tensor._make_subclass(cls, data)
 
+            __torch_function__ = torch._C._disabled_torch_function_impl
```
[the remark this answers is not in the document]
Comment (albanD) [reply test/test_autograd.py:7942]: Yes depending on which one lands first.

**Comment by albanD** [comment]
FYI @ezyang as well as you're thinking on how to replace this system altogether.

**Comment by albanD** [comment]
This is ready for a final review

**Review by soulitzer — APPROVED** [review]
LGTM

**Comment by zou3519** [comment]
@albanD @soulitzer this PR broke functorch: https://app.circleci.com/pipelines/github/pytorch/functorch/2032/workflows/2a0f2124-3870-4c6e-b95b-00951c1e9332/jobs/12098

I don't know the exact reason, but I suspect it has to do with the following:
1. functorch's DynamicLayerFrontMode is active inside of vmap.
2. PythonTLS saves DynamicLayerFrontMode as a part of the "saved TLS".
3. Somehow this goes wrong.

This doesn't seem easy to resolve. Could you help take a look please?

Minimal repro:
```
import torch
import functorch
from functorch import vmap, make_fx

def f(x):
    return torch.sin(x)
inp = torch.randn(5, 3)
f = vmap(f)
fx_f = make_fx(f)(inp)
new_inp = torch.randn(5, 3)
```

**Comment by albanD** [comment]
Hmmm not sure why this PR would change this behavior as this was happening before?
But yes, functorch breaks some of the assumptions we make here about how the dispatcher is used by jumping at specific keys. So I am not very surprised :/

**Comment by zou3519** [comment]
It might not look this way because of our lack of test coverage for this, but specifically this PR seems to break the torch_dispatch interaction with functorch. The main consequence of this right now is that we cannot compose AOTAutograd with functorch eager-mode transforms (e.g. vmap, grad, etc), so this is a really large breakage for us.

It would be a huge favor to us if we could figure out how to resolve this soon.

**Comment by zou3519** [comment]
Here's my current understanding to what is happening.

The purpose of this PR is to save the TLS when going from Python -> into the dispatcher so that when `__torch_dispatch__` is invoked, the TLS may be set to the saved one.

In the example,
```
f = vmap(torch.sin)
fx_f = make_fx(f)(inp)
```
we are:
- creating a ProxyTensor object, then calling vmap(f)(ProxyTensor)
- vmap adds DynamicLayerFrontMode to the dispatch local include key set.
- DynamicLayerFrontMode gets removed from the local include set after the batching rule for sin gets called.
- ProxyTensor.sin re-enables DynamicLayerFrontMode, which is a problem because we've already handled vmap.

So I think the question now is how should the TLS logic in this PR interact with mode-style dispatch keys?
- Let's say we have a HypotheticalMode dispatch key, and it is sandwiched between [PythonTLS, HypotheticalMode, Python].
- Furthermore, let's say that HypotheticalMode::sin (1) prints "sin" (2) temporarily removes HypotheticalMode from the local include set (in other words, it is "done"), and (3) dispatches down the line (either call or redispatch).

Imagine some Python code that looks like:
```
subclass = TensorSubclass(x)
add_to_tls_local_include_set(HypotheticalMode)
subclass.sin()
```
This would end up running through HypotheticalMode twice. Would that be the desired behavior?

**Comment by albanD** [comment]
> Would that be the desired behavior?

Yes. The high level idea of this TLS logic is the following: if you call `torch.foo(*args)` from top python or from within the torch_dispatch of `torch.foo` for one of the args. You should get the exact same result.

So in this case, that looks like the expected behavior to me? (not that we want to keep it necessarily but at least it does follow the high level idea)

**Pull request closed without merging**